File: TopHat_Workflow/se_alignment.pl
#Code written by Charles Warden (cwarden@coh.org, x60233)

use warnings;
use strict;

my $root = "/path/to/folder";
my $ref = "/path/to/ref";
my $threads = 4;

my $SE_reads_folder = "$root/Reads";

my $alignment_folder = "$root/TopHat_Alignment";

my @finished_files = ();
RNA_Seq_workflow_se_no_gz(\@finished_files, $SE_reads_folder, $alignment_folder, $threads, $ref);
exit;
	
sub RNA_Seq_workflow_se_no_gz
	{
		my ($arr_ref, $reads, $outputfolder, $threads, $ref)=@_;
		my @include_samples = @$arr_ref;
		
		#standard stuff
		
		opendir DH, $reads or die "Failed to open $reads: $!";
		my @files = readdir(DH);
		foreach my $file (@files)
			{
				#print "$file\n";
				if(-f ("$reads/$file") && ($file =~ /_L\d{3}_R1_001.fastq/))
					{
						my ($sampleID)= ($file =~ /(.*)_\w{6}_L\d{3}_R1_001.fastq/);
						print "Working on $sampleID\n";
										
						my $sample_folder = "$sampleID";
									
						my $new_flag = 1;
							
						foreach my $old_sample (@include_samples)
							{
								if ($old_sample eq $sample_folder)
										{
											$new_flag = 0;
										}
							}#end foreach my $old_sample (@exclude_samples)
								
						if($new_flag)
							{
								my $output_subfolder="$outputfolder/$sample_folder";
								mkdir($output_subfolder);
									
								my $read1 = "$reads/$file";
								
								print "\n\nAlign via TopHat\n\n";
								my $command = "tophat -o $output_subfolder -p $threads --no-coverage-search $ref $read1";
								system($command);
								
								my $bam_file = "$output_subfolder/accepted_hits.bam";																			
								my $rename_bam = "$outputfolder/$sampleID.bam";
								
								print "\n\nCreate Sorted BAM File\n\n";
								$command = "/opt/samtools-1.3/bin/samtools sort $bam_file -o $rename_bam";
								system($command);

								$command = "rm $bam_file";
								system($command);

								print "\n\nIndexing BAM File\n\n";
								$command = "/opt/samtools-1.3/bin/samtools index $rename_bam";
								system($command);
							}#end if if($new_flag)
					}#end if(-f ("$inputfolder/$file") && ($file =~ /_R1_/))
			}#end foreach my $file (@files)
		closedir(DH);
	}#end DNA_target_workflow

File: TopHat_Workflow/se_alignment.py
import sys
import re
import os

parameterFile = "parameters.txt"
finishedSamples = []

threads = ""
ref = ""
alignmentFolder = ""
readsFolder = ""
strandType = ""

inHandle = open(parameterFile)
lines = inHandle.readlines()
			
for line in lines:
	line = re.sub("\n","",line)
	line = re.sub("\r","",line)
	
	lineInfo = line.split("\t")
	param = lineInfo[0]
	value = lineInfo[1]

	if param == "Reads_Folder_MAC":
		readsFolder = value
	
	if param == "Alignment_Folder_MAC":
		alignmentFolder = value
		
	if param == "Bowtie2_Ref":
		ref = value
		
	if param == "Threads":
		threads = value

	if param == "strand":
		strandType = value

if (strandType == "") or (strandType == "[required]"):
	print "Need to enter a value for 'strand'!"
	sys.exit()
		
if (threads == "") or (threads == "[required]"):
	print "Need to enter a value for 'Threads'!"
	sys.exit()
	
if (ref == "") or (ref == "[required]"):
	print "Need to enter a value for 'Bowtie2_Ref'!"
	sys.exit()

if (readsFolder == "") or (readsFolder == "[required]"):
	print "Need to enter a value for 'Reads_Folder_MAC'!"
	sys.exit()
	
if (alignmentFolder == "") or (alignmentFolder == "[required]"):
	print "Need to enter a value for 'Alignment_Folder_MAC'!"
	sys.exit()
	
fileResults = os.listdir(readsFolder)

for file in fileResults:
	result = re.search("(.*)_S\d+_L\d{3}_R1_001.fastq$",file)
	
	if result:
		sample = result.group(1)
		if sample not in finishedSamples:
			print sample

			outputSubfolder = alignmentFolder +"/" + sample
			command = "mkdir " + outputSubfolder
			os.system(command)
									
			read1 = readsFolder + "/" + file
			
			tophatStrand = ""
			if strandType == "no":
				tophatStrand = "fr-unstranded"
			elif strandType == "reverse":
				tophatStrand = "fr-firststrand"
			elif strandType == "yes":
				tophatStrand = "fr-secondstrand"
			else:
				print "Need to provide TopHat mapping for strand: " + strandType
				sys.exit()
				
			print "\n\nAlign via TopHat\n\n"
			command = "tophat -o " + outputSubfolder + " -p " + threads + " --no-coverage-search --library-type " + tophatStrand + " " + ref + " " + read1 
			os.system(command)
								
			topHatBam = outputSubfolder + "/accepted_hits.bam";																			
			userBam = alignmentFolder + "/" + sample + ".bam";
								
			print "\n\nCreate Sorted BAM File\n\n"
			command = "/opt/samtools-1.4/samtools sort " + topHatBam + " -o " + userBam
			os.system(command)

			command = "rm " + topHatBam
			os.system(command)

			print "\n\nIndexing BAM File\n\n"
			command = "/opt/samtools-1.4/samtools index " + userBam
			os.system(command)
			
			print "\n\nCompressing .fastq File\n\n"
			command = "gzip " + read1
			os.system(command)
